Improvements and bug fixes for the TOF Converter window

jclarkeSTFC opened the pull request:
Some work on the TOF Converter window that's a combination of an improved layout, some refactoring, and some bug fixes. These changes fix a couple of issues (#33898 and #35585)

**Purpose of work**
<!-- Why has this work been done? If there is no linked issue please provide appropriate context for this work.
-->
Fixes #33898 and fixes #35585 

<!-- If the original issue was raised by a user they should be named here. Do not leak email addresses
**Report to:** [user name]
-->

**Summary of work**
<!-- Please provide a short, high level description of the work that was done.
-->
- Changed layout of window so the boxes are now centred.
- If the input and output units are the same then the convert button is disabled.
- The inputs for scattering angle and TOF are not cleared when changing units, unless you change to a unit that does not require those values.
- The scattering angle is normalised to [0, 360) degrees before use in the calculation, which prevents oddities like negative d-spacing noted in #33898.
- Disable scattering angle input when converting between d-spacing and momentum transfer, because in this case the angle is not required.
- Option added to take input value from any single-value workspace.

**To test:**

<!-- Instructions for testing.
There should be sufficient instructions for someone unfamiliar with the application to test - unless a specific
reviewer is requested.
If instructions for replicating the fault are contained in the linked issue then it is OK to refer back to these.
-->
- Check that unit conversions match previous version. Can also check that 1  meV = 1.6 × 10-22 J = 0.24 THz = 8.1 cm-1 = 11.6 K, which is from some notes.
- Try inputting angles outside the range [0, 360).
- Try converting between d-spacing and momentum transfer and check the scatting angle and TOF boxes are disabled.
- Try adding some single-value workspaces, some other workspaces, then check that the input option behaves correctly.

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
If you add release notes please save them as a separate file using the Issue or PR number as the file name. Check the file is located in the correct directory for your note(s).
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

#### Gatekeeper ####

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

jclarkeSTFC committed:
Use variables for the items in the two combo boxes, and add them in code, instead of in the .ui file.

jclarkeSTFC committed:
Rearranged layout of TOF converter window.

Made the window have centred boxes, in a fixed layout, instead of manually setting box sizes and location. Deleted some redundant text labels.

jclarkeSTFC committed:
Deleted try-except logic in imports.

This try-except was added in 2018, I think it was related to some refactoring. Doesn't seem required now unless I've missed something.

jclarkeSTFC committed:
Disable things if units are the same.

If the input and output units are the same, then disable several things in the interface.

jclarkeSTFC committed:
Don't clear TOF inputs unnecessarily.

Fixes #35585. The disabling of the boxes at the start of this method is not needed, instead it only disables them if not applicable to the current units. This also tidied up the logic further down the method.

jclarkeSTFC committed:
Normalise input angle to [0, 360)

The scattering angle should be in [0, 360), otherwise when converting to d-spacing you end up dividing by sin(angle > pi) which is going to give you a negative d-spacing.

jclarkeSTFC committed:
Disable angle for d-spacing to momentum transfer

When converting between d-spacing and momentum transfer the scattering angle is not required, although it is required when either of those units is used with another unit.

jclarkeSTFC committed:
Add release notes.

jclarkeSTFC committed:
Update TOF Converter docs.

jclarkeSTFC committed:
Add single-value workspaces as input option.

Add any single-value workspaces as options to a new combo box, so you can either pick one of those, or enter a specified value. It will remember the last specified value that you typed in when you switch input method.

jhaigh0 commented on qt/python/mantidqtinterfaces/mantidqtinterfaces/TofConverter/convertUnits.py:
```diff
@@ -11,6 +11,12 @@
 
 
 def doConversion(inputval, inOption, outOption, theta, flightpath):
+    if (
+        (inOption == MOMENTUM_TRANSFER or inOption == D_SPACING)
+        and (outOption == MOMENTUM_TRANSFER or outOption == D_SPACING)
+        and inOption != outOption
+    ):
```
I think this is equivalent?

```suggestion
    if (inOption == MOMENTUM_TRANSFER and outOption == D_SPACING)
        or (inOption == D_SPACING and outOption == MOMENTUM_TRANSFER):
```

jhaigh0 reviewed (changes requested):
Some initial code changes for you.

jclarkeSTFC committed:
Update release notes.

Comment about using single valued workspaces was missing.

jclarkeSTFC committed:
Move constants to top of file.

jclarkeSTFC committed:
Fix some logic and reduce duplication.

Fixed logic for determining whether it's a momentum transfer to d-spacing conversion or not, and put the logic in one place only.

jclarkeSTFC committed:
Add comment.

jhaigh0 reviewed (approved):
Thanks for the changes 👍

gemmaguest commented:
@jclarkeSTFC In general I'd prefer see separate PRs rather than one larger PR containing multiple changes like this. Don't be afraid of creating tiny PRs! In general, the smaller the better :) I know that can feel like unnecessary overhead, but I'd rather focus on reducing the overhead of getting PRs through if that is the case! Advantages of smaller PRs include:

- they are (usually) easier to review;
- you will get better feedback;
- they usually get merged more quickly;
- if there are any reviewer changes to address with one change it doesn't hold up the other changes;
- the history of why a change was made is clearer;
- if a problem is introduced it can be tracked down more easily;
- if necessary, one change can be reverted without affecting other changes.

Of course it might make sense to do them together for some reason, e.g. they make sense as a single release note or make life easier for the tester - it's perhaps worth mentioning it in the PR description if there was a deliberate decision along those lines.
